### PR TITLE
Change completely transactions table

### PR DESCRIPTION
**Pull request by tdeo**
@matthieugendreau, this changes completely the transactions table to stop storing duplicated rows, and also adds a AUTo INCREMENT PK

### CODE REVIEW

**Comment by tdeo** [comment]
This will require running `DELETE FROM transactions WHERE price < 0 OR (price = 0 AND buyer_id < receiver_id)` to remove duplicates

File: app/assets/javascripts/transaction.js [hunk 1]
```diff
@@ -2,10 +2,7 @@ function buttonCancelTransation() {
   $('.cancel-transaction').click(function(event) {
     $.ajax({
       data: {
-        id: event.target.parentElement.getAttribute('t-id'),
-        id2: event.target.parentElement.getAttribute('t-id2'),
-        date: event.target.parentElement.getAttribute('t-date'),
-        price: event.target.parentElement.getAttribute('t-price'),
+        transaction_id: event.target.parentElement.getAttribute('t-id'),
```
Comment (matthieu-gendreau) [inline app/assets/javascripts/transaction.js:5]: a bit confusing to translate use transaction_id on one side and t-id on the other side.
I would prefer t-transaction-id to be sure of what we manipulate during the transition

File: db/migrate/20170621181432_add_pk_to_transactions.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  def change
+    rename_column :transactions, :id, :buyer_id
+    rename_column :transactions, :id2, :receiver_id
+    rename_column :transactions, :price, :amount
+
+    reversible do |dir|
+      dir.up do
+        ActiveRecord::Base.connection.execute 'ALTER TABLE `transactions` ADD `id` INT NOT NULL AUTO_INCREMENT PRIMARY KEY FIRST'
+        ActiveRecord::Base.connection.execute 'ALTER TABLE `transactions` ENGINE=InnoDB'
+        ActiveRecord::Base.connection.execute 'ALTER TABLE `transactions` CONVERT TO CHARACTER SET utf8mb4'
```
Comment (matthieu-gendreau) [inline db/migrate/20170621181432_add_pk_to_transactions.rb:11]: From what I understand i includes more characters than utf8 (🍵🍶🍷🍸).
Is there a specific reason though not to use utf8?

File: app/controllers/account_controller.rb [hunk 3]
```diff
@@ -184,12 +185,10 @@ def filter
...
 
   def cancel_transaction
-    date = Time.zone.parse(params[:date])
-    transaction = Transaction.find_by(id: params[:id], id2: params[:id2], price: params[:price], date: date)
-    reverse = Transaction.find_by(id: params[:id2], id2: params[:id], price: -params[:price].to_i, date: date)
-    fail TdbException, 'Cannot find this transaction' unless transaction && reverse
-    require_admin!(:credit) if date < Time.current - 1.minute
-    transaction.cancel(reverse, @admin)
+    transaction = Transaction.find_by(id: params[:transaction_id])
+    fail TdbException, 'Cannot find this transaction' unless transaction
```
Comment (matthieu-gendreau) [inline app/controllers/account_controller.rb:189]: I18n.t('find_transaction_fail') :stuck_out_tongue:

File: db/migrate/20170621180448_add_cigarettes_to_transactions.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class AddCigarettesToTransactions < ActiveRecord::Migration
+  def change
+    add_column :transactions, :is_tobacco, :boolean, null: false, default: false, after: :id2
```
Comment (matthieu-gendreau) [inline db/migrate/20170621180448_add_cigarettes_to_transactions.rb:3]: add_column :transactions, :is_cancer, :boolean, null: false, default: false, after: :id2
:smoking:

File: app/models/transaction.rb [hunk 5]
```diff
@@ -1,24 +1,35 @@
...
       amounts.each do |account, amount|
-        amount = (amount * 100).round(1).ceil # Screwing the customers
-        account.balance = account.balance - amount
-        account.turnover = account.turnover + amount if amount > 0
-        bank.balance = bank.balance + amount
-        bank.turnover = bank.turnover - amount if amount < 0
-        create(id: account.id, id2: bank.id, price: -amount, comment: comment, admin: admin.try(:id), date: time)
-        create(id: bank.id, id2: account.id, price: amount, comment: comment, admin: admin.try(:id), date: time)
+        buyer = amount > 0 ? account : bank
+        receiver = amount > 0 ? bank : account
```
Comment (matthieu-gendreau) [inline app/models/transaction.rb:16]: if amount > 0 then buyer, receiver = account, bank else buyer, receiver = bank, account end

File: app/models/transaction.rb [hunk 6]
```diff
@@ -1,24 +1,35 @@
...
+        amount = -amount if amount < 0
```
Comment (matthieu-gendreau) [inline app/models/transaction.rb:17]: amount = amount.abs

Comment (matthieu-gendreau) [inline app/models/transaction.rb:17]: value = amount.abs
I am a bit reluctant to use the same variable name the describe something different

File: app/views/account/_show_description.html.erb [hunk 7]
```diff
@@ -34,3 +34,6 @@
   <br>
 </div>
 <h1 class="<%= 'text-danger' if @account.balance < 0 %>"><strong><%= number_to_currency(@account.balance / 100.0) %></strong></h1>
+<% if @account.total_clopes > 0 %>
+  <h4><strong><%= number_to_currency(@account.total_clopes / 100.0) %></strong> dépensés en cigarettes</h4>
+<% end %>
```
Comment (matthieu-gendreau) [inline app/views/account/_show_description.html.erb:39]: Not sure if this is really relevant for anyone

File: app/views/transactions/_transaction.html.erb [hunk 8]
```diff
@@ -1,21 +1,32 @@
 <tr>
+  <%
+    other_account, amount = nil, nil
+    is_debit = transaction.buyer_id == @account.id
+    if is_debit
+      other_account = transaction.receiver
+    else
+      other_account = transaction.buyer
```
Comment (matthieu-gendreau) [inline app/views/transactions/_transaction.html.erb:8]: other_account = is_debit ? transaction.receiver : transaction.buyer

File: app/views/transactions/_transaction.html.erb [hunk 9]
```diff
@@ -1,21 +1,32 @@
 <tr>
+  <%
+    other_account, amount = nil, nil
```
Comment (matthieu-gendreau) [inline app/views/transactions/_transaction.html.erb:3]: Why this line?

Comment (matthieu-gendreau) [inline app/views/transactions/_transaction.html.erb:3]: I don't like the name other_account but I can't come up with something better :confused:

File: config/locales/fr.yml [hunk 10]
```diff
@@ -14,6 +14,8 @@ fr:
   password_again: Mot de passe (confirmation)
   pay: Paye, bâtard !
   price: Prix
+  debited: Débit
```
Comment (matthieu-gendreau) [inline config/locales/fr.yml:17]: debit: Débit
debited: Débité
No?

**Review by matthieu-gendreau — APPROVED** [review]
:+1:

File: config/locales/fr.yml (hunk 10 above)
Comment (tdeo) [reply config/locales/fr.yml:17]: We already had a `debit` key to "Débiter".

File: app/views/transactions/_transaction.html.erb (hunk 9 above)
Comment (tdeo) [reply app/views/transactions/_transaction.html.erb:3]: Can we call that the `partner` as we're in the context of a transaction ?

File: db/migrate/20170621181432_add_pk_to_transactions.rb (hunk 2 above)
Comment (tdeo) [reply db/migrate/20170621181432_add_pk_to_transactions.rb:11]: utf8 only supports 3-byte characters maximum, which is not true utf8

**Comment by tdeo** [comment]
Comments addressed in https://github.com/bobar/tdfb/commit/2461c1469fd20e634c501e17fdea7baa5200ad2c

**Pull request merged**